src/Http/Paginators/OffsetRequestPaginator.php

```diff
@@ -0,0 +1,294 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Saloon\Http\Paginators;
+
+use Saloon\Contracts\Connector;
+use Saloon\Contracts\Request;
+
+// TODO 1: Look into serialising the Connector and original Request,
+//           to ensure that we can rebuild the paginator state without storing the entire multiverse.
+// TODO 2: Make it easier to extend the RequestPaginator. Preferably via callbacks, so we ideally don't even need separate classes.
+// TODO 3: Because both page-based and offset-based pagination just bumps the 'paging' number,
+//           would it make sense to do all pagination in _the_ RequestPaginator, but allow to specify the counter somehow?
+//         Maybe a callback that receives a copy of the original Request, as well as the latest Response (which has the corresponding latest Request and PendingRequest),
+//           and have that callback set the next 'page' on the new Request?
+
+/**
+ * @template TRequest of \Saloon\Contracts\Request
+ * @template TResponse of \Saloon\Contracts\Response
+ *
+ * @extends RequestPaginator<TRequest, TResponse>
+ *
+ * @method int limit()
+ */
+class OffsetRequestPaginator extends RequestPaginator
+{
+    /**
+     * @var int
+     */
+    protected readonly int $originalOffset;
+
+    /**
+     * @param \Saloon\Contracts\Connector $connector
+     * @param \Saloon\Contracts\Request $originalRequest
+     * @param int<0, max> $limit
+     * @param int<0, max> $offset
+     */
+    public function __construct(
+        Connector $connector,
+        Request $originalRequest,
+        int $limit,
+        protected int $offset = 0,
+    ) {
+        parent::__construct($connector, $originalRequest, $limit);
+
+        $this->originalOffset = $offset;
+    }
+
+    /**
+     * @return int
+     */
+    public function totalEntries(): int
+    {
+        // Make sure we have a response.
+        if (is_null($this->currentResponse)) {
+            $this->current();
+        }
+
+        return $this->currentResponse->json('total');
+    }
+
+    /**
+     * @return iterable<int, array<string, mixed>>
+     *
+     * @TODO entry data type
+     */
+    public function entries(): iterable
+    {
+        return $this->currentResponse->json('data');
+    }
+
+    /**
+     * @return int
+     */
+    public function totalPages(): int
+    {
+        // Make sure we have a response.
+        if (is_null($this->currentResponse)) {
+            $this->current();
+        }
+
+        return $this->lastPage();
+    }
+
+    /**
+     * @return int
+     */
+    public function firstPage(): int
+    {
+        // TODO: Or should we use something from the response?
+        return 1;
```

Comment:
I think this is fine, but allow people to extend it, which looks like people can